OSDOCS-18134: JTBD Docs Enhancement Gateway API Routing (#109001)
openshift/openshift-docs: modules/applying-processing-filters-http-requests.adoc

`requestMirror`::
Available on {SMProductName}, this filter mirrors, or duplicates, requests to a specified `backendRef`. You can optionally specify a `percent` or `fraction` field to dictate the portion of requests to mirror. If omitted, 100% of the requests are mirrored.

`urlRewrite`::
Available on {SMProductName}, this filter modifies the hostname, path, or both for an HTTP request. It uses the same path semantics as the `requestRedirect` filter. 

.Prerequisites

* You have access to the cluster as a user with the `cluster-admin` role.
* You have installed the {oc-first}.

.Procedure

. Create or edit an `HTTPRoute` YAML file to include your desired processing directives under the `spec.rules.filters` field. 
+
--
The following example demonstrates how to configure a `requestRedirect` filter that issues a permanent redirect (301) from HTTP to HTTPS:

[source,yaml]
----
spec:
  hostnames:
  - redirect.mycompany.com
  rules:
  - filters:
    - type: RequestRedirect
      requestRedirect:
        scheme: https
        statusCode: 301

> ----

Reviewer: would suggest placing a full top-down YAML file to give readers the idea of a HTTPRoute

// Module included in the following assemblies:
//

> // * routing-http-requests-to-services.adoc

Reviewer 2: ```suggestion
// * networking/ingress_load_balancing/configuring_ingress_cluster_traffic/routing-http-requests-to-services.adoc
```
metadata should contain the whole path

//
:_mod-docs-content-type: PROCEDURE
[id="applying-processing-filters-http-requests_{context}"]
= Applying processing filters to HTTP requests

[role="_abstract"]

> When a request hits your route, you can apply processing directions to the HTTP request, such as modifying headers or redirecting to another URL, by configuring filters. You can specify up to 16 filters for a single rule. 

Reviewer 2: ```suggestion
You can apply processing directions to HTTP requests on a specified route by configuring filters. For example, you can modify headers or redirect traffic to another URL. You can specify up to 16 filters for a single rule. 
```
What's a rule? This is the only mention in the entire procedure of "rule." Do you mean route?
Why would we do this? Why is an essential ingredient in JTBD.
Also, I don't think the flow here is right. We aren't waiting until requests come in and then configuring rules on the fly, which is implied in the way this is written. We're adding these filters beforehand as part of our route config, right?

> You can generally combine multiple filters to achieve advanced processing results. However, the `urlRewrite` and `requestRedirect` filters are mutually exclusive and cannot be used in combination.

Reviewer 2: Lines 12-29 do not belong in a procedure. This is all reference and should be modularized accordingly. Suggest creating a reference module and including after this procedure.

`requestHeaderModifier`::
Modifies the headers of an HTTP request. You can specify only one modifier per header, and multiple values for a single header must be comma-separated. You can list up to 16 header filters using the `set`, `add`, or `remove` fields.

`responseHeaderModifier`::
Available on {SMProductName}, this filter modifies the headers of an HTTP response. It uses the exact same constraints and fields as the `requestHeaderModifier` filter.

`requestMirror`::
Available on {SMProductName}, this filter mirrors, or duplicates, requests to a specified `backendRef`. You can optionally specify a `percent` or `fraction` field to dictate the portion of requests to mirror. If omitted, 100% of the requests are mirrored.

`urlRewrite`::
Available on {SMProductName}, this filter modifies the hostname, path, or both for an HTTP request. It uses the same path semantics as the `requestRedirect` filter. 

.Prerequisites

* You have access to the cluster as a user with the `cluster-admin` role.
* You have installed the {oc-first}.

.Procedure

> . Create or edit an `HTTPRoute` YAML file to include your desired processing directives under the `spec.rules.filters` field. 

Reviewer 2: i don't think this is the way to go here; you are trying to be general and specific in the same step; i would just give one example and refer readers to a reference module

+
--

> The following example demonstrates a complete `HTTPRoute` resource with a `requestRedirect` filter that issues a permanent redirect (301) from HTTP to HTTPS:

Reviewer 2: ```suggestion
The following example demonstrates a complete `HTTPRoute` custom resource (CR) with a `requestRedirect` filter that issues a permanent redirect from HTTP to HTTPS:
```
specify resource type always (CR or CRD)

[source,yaml]
----
apiVersion: gateway.networking.k8s.io/v1
kind: HTTPRoute
metadata:
  name: http-filter-example

>   namespace: my-application

Reviewer 2: ```suggestion
  namespace: <example_application>
```
we don't use "my" in user-replaceable values
use underscores and angle brackets for all user-replaceable values

spec:
  parentRefs:
  - name: my-gateway
    namespace: openshift-ingress
  hostnames:
  - "example.com"
  rules:
  - filters:
    - type: RequestRedirect
      requestRedirect:
        scheme: https
        statusCode: 301
----
* `hostnames` defines the domain, such as `"example.com"`, that this route applies to.
* `filters` specifies the processing logic. In this example, the `RequestRedirect` type is used.
* `scheme: https` instructs the gateway to redirect the client to the secure version of the URL.
* `statusCode: 301` indicates a permanent redirect.

> The following snippet demonstrates how to configure a `requestHeaderModifier` filter that adds a new header, modifies an existing header, and removes a specific header:

Reviewer 2: this content is reference, not procedure
